transactions/btc.ts

```diff
@@ -526,6 +527,201 @@ export async function getBtcFeesForNonOrdinalBtcSend(
+
+    tx.sign(hex.decode(dummyPrivateKey));
+    tx.finalize();
+
+    const txSize = tx.vsize;
+    let sentSats = new BigNumber(0);
+
+    for (let i = 0; i < tx.outputsLength; i++) {
+      const output = tx.getOutput(i);
+      sentSats = sentSats.plus(new BigNumber((output.amount ?? 0n).toString()));
+    }
+
+    const change = sentSats.minus(recipientTotal);
+    const fee = new BigNumber(txSize).times(feeRate);
+
+    // check if there is change and if the change is greater than the vsize*feeRate + dust rate
+    if (change.gt(fee.plus(BITCOIN_DUST_VALUE))) {
+      const changeWithoutFee = change.minus(fee);
+      return {
+        selectedUtxos,
+        fee: fee.toNumber(),
+        feeRate: feeRate,
+        change: changeWithoutFee.toNumber(),
+      };
+    }
+  }
+
+  {
+    // try txn without change
+    const txNoChange = createTransaction(
+      dummyPrivateKey,
+      selectedUtxos,
+      recipientTotal,
+      recipients,
+      changeAddress,
+      'Mainnet',
+      true,
+    );
+
+    txNoChange.sign(hex.decode(dummyPrivateKey));
+    txNoChange.finalize();
+
+    const txSize = txNoChange.vsize;
+    let sentSats = new BigNumber(0);
+
+    for (let i = 0; i < txNoChange.outputsLength; i++) {
+      const output = txNoChange.getOutput(i);
+      sentSats = sentSats.plus(new BigNumber((output.amount ?? 0n).toString()));
+    }
+
+    const fee = inputSum.minus(sentSats);
+
+    if (fee.div(txSize).gt(feeRate)) {
+      return {
+        selectedUtxos,
+        fee: fee.toNumber(),
+        feeRate: fee.div(txSize).toNumber(),
+        change: 0,
+      };
+    }
+  }
+
+  return undefined;
+}
+
+function selectOptimalUtxos(
+  recipients: Recipient[],
+  selectedUtxos: UTXO[],
+  availableUtxos: UTXO[],
+  changeAddress: string,
+  feeRate: number,
+  currentBestUtxoCount: number | undefined = undefined,
+): TransactionUtxoSelectionMetadata | undefined {
+  const currentSelectionData = getTransactionMetadataForUtxos(recipients, selectedUtxos, changeAddress, feeRate);
+
+  // if there is a valid selection, adding more UTXOs would only make the fees higher, so just return
+  if (currentSelectionData) return currentSelectionData;
+
+  // if we have a current best selection, we want to check if adding another UTXO would make the fees higher
+  // and skip if it does since it would be a worse selection
+  const wouldIncreaseFees = currentBestUtxoCount === selectedUtxos.length - 1;
+
+  if (availableUtxos.length === 0 || wouldIncreaseFees) {
+    // we either have no more UTXOs to add or adding would make an existing outer selection worse, so we bail
+    return undefined;
+  }
+
+  // sort smallest to biggest so we can pop biggest from end
+  const sortedUtxos = [...availableUtxos];
+  sortedUtxos.sort((a, b) => a.value - b.value);
+
+  let bestSelectionData: TransactionUtxoSelectionMetadata | undefined = undefined;
+  while (sortedUtxos.length > 0) {
+    const utxo = sortedUtxos.pop() as UTXO;
```

Comment:
```suggestion
    const utxo = sortedUtxos.pop();
```
prefer typing on L646 to casting, as casting will just ignore type errors

Reply:
This is on purpose. Maybe I should add comment 🤔 The sortedUtxos is a UTXO[] but the pop function could return undefined. Since the while loop checks that there are elements in the array, it is safe to cast it to UTXO. TS is fine with it, to completely bypass it you'd need to cast to unknown first `const utxo = sortedUtxos.pop() as unknown as UTXO`. 

Another option is to explicitly check that `utxo` is not undefined and throw an error if it is, though that should never happen due to the while loop check.

Reply:
I added a comment in the code. Let me know if you think it's fine or if it should maybe be an explicit not undefined check 🙏